metadata/report/delegate/delegate_report.go

```diff
@@ -0,0 +1,275 @@
+	}
+	return bmr, nil
+}
+
+// retry will do metadata failed reports collection by call metadata report sdk
+func (bmr *MetadataReport) retry() bool {
+	bmr.failedReportsLock.RLock()
+	defer bmr.failedReportsLock.RUnlock()
+	return bmr.doHandlerMetadataCollection(bmr.failedReports)
+}
+
+// StoreProviderMetadata will delegate to call remote metadata's sdk to store provider service definition
+func (bmr *MetadataReport) StoreProviderMetadata(identifier *identifier.MetadataIdentifier, definer definition.ServiceDefiner) {
+	if bmr.syncReport {
+		bmr.storeMetadataTask(common.PROVIDER, identifier, definer)
+	}
+	go bmr.storeMetadataTask(common.PROVIDER, identifier, definer)
+}
+
+// storeMetadataTask will delegate to call remote metadata's sdk to store
+func (bmr *MetadataReport) storeMetadataTask(role int, identifier *identifier.MetadataIdentifier, definer interface{}) {
+	logger.Infof("store provider metadata. Identifier :%v ; definition: %v .", identifier, definer)
+	bmr.allMetadataReportsLock.Lock()
+	bmr.allMetadataReports[identifier] = definer
+	bmr.allMetadataReportsLock.Unlock()
+
+	bmr.failedReportsLock.Lock()
+	delete(bmr.failedReports, identifier)
+	bmr.failedReportsLock.Unlock()
+	// data is store the json marshaled definition
+	var (
+		data []byte
+		err  error
+	)
+
+	defer func() {
+		if r := recover(); r != nil {
+			bmr.failedReportsLock.Lock()
+			bmr.failedReports[identifier] = definer
+			bmr.failedReportsLock.Unlock()
+			bmr.metadataReportRetry.startRetryTask()
+			logger.Errorf("Failed to put provider metadata %v in %v, cause: %v", identifier, string(data), r)
+		}
+	}()
+
+	data, err = json.Marshal(definer)
+	if err != nil {
+		logger.Errorf("storeProviderMetadataTask error in stage json.Marshal, msg is %v", err)
+		panic(err)
+	}
+	report := instance.GetMetadataReportInstance()
+	if role == common.PROVIDER {
+		err = report.StoreProviderMetadata(identifier, string(data))
+	} else if role == common.CONSUMER {
+		err = report.StoreConsumerMetadata(identifier, string(data))
+	}
+
+	if err != nil {
+		logger.Errorf("storeProviderMetadataTask error in stage call  metadata report to StoreProviderMetadata, msg is %v", err)
+		panic(err)
+	}
+}
+
+// StoreConsumerMetadata will delegate to call remote metadata's sdk to store consumer side service definition
+func (bmr *MetadataReport) StoreConsumerMetadata(identifier *identifier.MetadataIdentifier, definer map[string]string) {
+	if bmr.syncReport {
+		bmr.storeMetadataTask(common.CONSUMER, identifier, definer)
+	}
+	go bmr.storeMetadataTask(common.CONSUMER, identifier, definer)
+}
+
+// SaveServiceMetadata will delegate to call remote metadata's sdk to save service metadata
+func (bmr *MetadataReport) SaveServiceMetadata(identifier *identifier.ServiceMetadataIdentifier, url common.URL) error {
+	report := instance.GetMetadataReportInstance()
+	if bmr.syncReport {
+		return report.SaveServiceMetadata(identifier, url)
+	}
+	go report.SaveServiceMetadata(identifier, url)
```

Comment:
I think the o should be handler by the func called. Because we don't know what should we do because `SaveServiceMetadata` method is provided by extension metadata like ZooKeeper, etcd and so on. They should provider their error handler method.